packages/plugin-ext/src/hosted/node/plugin-host-proxy.ts

```diff
@@ -0,0 +1,80 @@
+/********************************************************************************
+ * Copyright (C) 2022 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as http from 'http';
+import * as https from 'https';
+import * as tls from 'tls';
+
+import { createHttpPatch, createProxyResolver, createTlsPatch, ProxySupportSetting } from 'vscode-proxy-agent';
+import { PreferenceRegistryExtImpl } from '../../plugin/preference-registry';
+
+export function connectProxyResolver(configProvider: PreferenceRegistryExtImpl): void {
+    const resolveProxy = createProxyResolver({
+        resolveProxy: async url => url,
+        getHttpProxySetting: () => configProvider.getConfiguration('http').get('proxy'),
+        log: () => { },
+        getLogLevel: () => 0,
+        proxyResolveTelemetry: () => { },
+        useHostProxy: true,
+        env: process.env,
+    });
+    const lookup = createPatchedModules(configProvider, resolveProxy);
+    return configureModuleLoading(lookup);
+}
+
+interface PatchedModules {
+    http: typeof http;
+    https: typeof https;
+    tls: typeof tls;
+}
+
+function createPatchedModules(configProvider: PreferenceRegistryExtImpl, resolveProxy: ReturnType<typeof createProxyResolver>): PatchedModules {
+    const proxySetting = {
+        config: 'off' as ProxySupportSetting
+    };
+    const certSetting = {
+        config: false
+    };
+    configProvider.onDidChangeConfiguration(() => {
+        const httpConfig = configProvider.getConfiguration('http');
+        proxySetting.config = httpConfig?.get<ProxySupportSetting>('proxySupport') || 'off';
+        certSetting.config = !!httpConfig?.get<boolean>('systemCertificates');
+    });
+
+    return {
+        http: Object.assign(http, createHttpPatch(http, resolveProxy, proxySetting, certSetting, true)),
+        https: Object.assign(https, createHttpPatch(https, resolveProxy, proxySetting, certSetting, true)),
+        tls: Object.assign(tls, createTlsPatch(tls))
+    };
+}
+
+function configureModuleLoading(lookup: PatchedModules): void {
```

Comment:
We already have code that overrides `_load` in `plugin-vscode-init` and `backend-init-theia`. It should be safe, because whichever one happens to go first will become the other's `original` and whichever goes second should delegate to that and both should get to do their work, but it still makes me a bit queasy. Would it be possible to do this load override in the same place as the other?

Reply:
I think this will lead to convoluted code, due to two issues, mainly because we don't have the `PreferenceRegistryExtImpl` easily available in the scope the `plugin-vscode-init` and `backend-init-theia` functions are running in. Trying to configure the proxy support at the place the `vscode`/`theia` load gets overriden seems to be therefore quite difficult.

I get the queasyness, but I think that this is the least convoluted way to achieve this.

Reply:
Sounds reasonable. Perhaps for a later refactoring of the plugin initialization system.